R/aes-evaluation.r

```diff
@@ -2,39 +2,106 @@
+#'   should use the `after_stat()` function to flag that evaluation of the
+#'   aesthetic mapping should be postponed until after stat transformation.
+#'   Evaluation after stat transformation will have access to the variables
+#'   calculated by the stat, not the original mapped values. The 'computed
+#'   variables' section in each stat lists which variables are available to
+#'   access.
+#'
+#' ```{r after_stat_transformation, eval = FALSE}
+#' # The 'y' values for the histogram are computed by the stat
+#' ggplot(faithful, aes(x = waiting)) +
+#'   geom_histogram()
+#'
+#' # Choosing a different computed variable to display, matching up the
+#' # histogram with the density plot
+#' ggplot(faithful, aes(x = waiting)) +
+#'   geom_histogram(aes(y = after_stat(density))) +
+#'   geom_density()
+#' ```
+#'
+#' ## Stage 3: after scale transformation
+#'   The third and last stage is after the data has been transformed and
+#'   mapped by the plot scales. An example of mapping from scaled data could
+#'   be to use a desaturated version of the stroke colour for fill. You should
+#'   use `after_scale()` to flag evaluation of mapping for after data has been
+#'   scaled. Evaluation after scaling will only have access to the final
+#'   aesthetics of the layer (including non-mapped, default aesthetics).
+#'
+#' ```{r after_scale_transformation, eval = FALSE}
+#' # The exact colour is known after scale transformation
+#' ggplot(mpg, aes(cty, colour = factor(cyl))) +
+#'   geom_density()
+#'
+#' # We re-use colour properties for the fill without a separate fill scale
+#' ggplot(mpg, aes(cty, colour = factor(cyl))) +
+#'   geom_density(aes(fill = after_scale(alpha(colour, 0.3))))
+#' ```
 #'
-#' @param x An aesthetic expression using variables calculated by the stat
-#'   (`after_stat()`) or layer aesthetics (`after_scale()`).
-#' @param start An aesthetic expression using variables from the layer data.
-#' @param after_stat An aesthetic expression using variables calculated by the
-#'   stat.
-#' @param after_scale An aesthetic expression using layer aesthetics.
+#' ## Complex staging
+#'   If you want to map the same aesthetic multiple times, e.g. map `x` to a
+#'   data column for the stat, but remap it for the geom, you can use the
+#'   `stage()` function to collect multiple mappings.
+#'
+#' ```{r complex_staging, eval = FALSE}
+#' # Use stage to modify the scaled fill
+#' ggplot(mpg, aes(class, hwy)) +
+#'   geom_boxplot(aes(fill = stage(class, after_scale = alpha(fill, 0.4))))
+#'
+#' # Using data for computing summary, but placing label elsewhere, as well as
+#' # using the computed 'y' as part of the label.
+#' ggplot(mpg, aes(class, displ)) +
+#'   geom_violin() +
+#'   stat_summary(
+#'     aes(y = stage(displ, after_stat = 8),
+#'         label = after_stat(paste(y, ymax, sep = " ± "))),
+#'     geom = "text",
+#'     fun = ~ round(mean(.x), 2),
+#'     fun.max = ~ round(sd(.x), 2)
```

Comment:
Small nitpicking only because of my love/hate relationship with the `fun.*` arguments in `stat_summary()`:

May I suggest just passing a function to `fun.data` returning a dataframe with the desired variables-as-columns, as opposed to supplying functions to `fun` and `fun.max` that each return vectors?

So those 2 lines can be replaced with:

```r
# possibly `\(val) ...` so the argument is not confused with the x-positional aesthetic
fun.data = \(x) round(data.frame(mean = mean(x), sd = sd(x)), 2)
```

Then, the columns from `fun.data` output can be referenced in the `after_stat()` for `label` like this, which IMO is more transparent:

```r
after_stat(paste(mean, sd, sep = " ± "))
```

And it also produces a more straightforward layer data:

```r
p_new <- ggplot(mpg, aes(class, displ)) +
   geom_violin() +
   stat_summary(
     aes(y = stage(displ, after_stat = 8),
         label = after_stat(paste(mean, sd, sep = " ± "))),
     geom = "text",
     fun.data = \(x) round(data.frame(mean = mean(x), sd = sd(x)), 2)
   )
layer_data(last_plot(), 2)[,1:7]
#>   y       label x group mean   sd PANEL
#> 1 8 6.16 ± 0.53 1     1 6.16 0.53     1
#> 2 8 2.33 ± 0.45 2     2 2.33 0.45     1
#> 3 8 2.92 ± 0.72 3     3 2.92 0.72     1
#> 4 8 3.39 ± 0.45 4     4 3.39 0.45     1
#> 5 8 4.42 ± 0.83 5     5 4.42 0.83     1
#> 6 8  2.66 ± 1.1 6     6 2.66 1.10     1
#> 7 8 4.46 ± 1.07 7     7 4.46 1.07     1
```

Without overriding `fun.data` (or doing an ugly `fun.min = ~ NULL`), you get the vestigial `ymin` column that's there to support the default pointrange geom. Plus, the mean that's calculated from `fun` gets mapped to `y` and then is immediately overriden by `stage(after_stat = 8)` which may be surprising (in case one wanted to use it to map `size` in the after-scale of something):

```r
p_old <- ggplot(mpg, aes(class, displ)) +
   geom_violin() +
   stat_summary(
     aes(y = stage(displ, after_stat = 8),
         label = after_stat(paste(y, ymax, sep = " ± "))),
     geom = "text",
     fun = ~ round(mean(.x), 2),
     fun.max = ~ round(sd(.x), 2)
   )
layer_data(last_plot(), 2)[,1:7]
#>   y       label x group ymin ymax PANEL
#> 1 8 6.16 ± 0.53 1     1   NA 0.53     1
#> 2 8 2.33 ± 0.45 2     2   NA 0.45     1
#> 3 8 2.92 ± 0.72 3     3   NA 0.72     1
#> 4 8 3.39 ± 0.45 4     4   NA 0.45     1
#> 5 8 4.42 ± 0.83 5     5   NA 0.83     1
#> 6 8  2.66 ± 1.1 6     6   NA 1.10     1
#> 7 8 4.46 ± 1.07 7     7   NA 1.07     1
```

Lastly and relatedly, because sd is held in the `ymax` column in the old example, it gets used later in the retraining of the y-scales, so if you plot `p_old` you see it actually gets an expansion down to around 0 in the y-axis because `ymax` is around that value. All of this mess because `stat_summary()` appears as if it's geom-agnostic when it kinda isn't 😅

I may be missing obvious problems with my approach and this is possibly off-label usage of `fun.data`, but as long as `fun.data` dataframe doesn't touch standard aesthetics (like overriding/duplicating the `x` column or hard-coding `colour` before the scale sees it), this should be ok?

(And of course thanks for this great addition to the docs!)

Reply:
Thanks June, I love the suggestion. It makes it much clearer about what is going on! Especially the y-axis expansion I hadn't spotted, so thanks :) The only way I'll deviate from your suggestion is that the example should work in older R versions than R4.1.0, so I'll stick to the rlang lambda syntax instead of base R lambda syntax for now.